### ISSUE TITLE
TypeScript Intellisense extremely lagged after upgrading to 0.10.10

### ISSUE THREAD

**Issue by bill-mybiz** [Microsoft/vscode#3958]
After upgrading to 0.10.10, my Intellisense is now extremely inconsistent and laggy. It happens with most auto-complete attempts, but is most easily reproduced by this [excerpt from another comment I made](https://github.com/Microsoft/vscode/issues/3791#issuecomment-194300157):

> When I hit ctrl+space on "trycatch" to get the snippet intellisense option, it will consistently lag about 10sec. Ctrl+space thereafter will be immediate as long as I don't edit anything. But if I hit spacebar just after trycatch (to move the cursor off of the symbol), then left arrow (to return to the end of "trycatch"), and then ctrl+space, there will be the 10 second lag again.

There are some possibly related issues that might help. At first I thought it was related to [this javascript issue](https://github.com/Microsoft/vscode/issues/3791), but it appears not to be the case. Since it is only lagged and not completely failed, I don't think it is related to [this Typescript issue](https://github.com/Microsoft/vscode/issues/3889), but it could be. It also resembles [this other Format Code issue](https://github.com/Microsoft/vscode/issues/3652), but again in my case it does not completely fail. Perhaps because [my project](https://github.com/ibgib/stackGib) is not as large and theirs is possibly just extremely lagged to the point of crashing. 

@egamma suggests in [this comment](https://github.com/Microsoft/vscode/issues/3791#issuecomment-194774281) that is the files glob that may be the culprit. We're going to look into this further as I was not aware it was considered an Atom-specific practice, but I would think it unlikely to be the problem as it worked previously without problems and that there are multiple other issues regarding Intellisense problems after the upgrade.

**Comment by bill-mybiz** [comment]
Ah, with regards to the files glob, we do not use the Atom-specific implementation. As per [this TS files glob issue](https://github.com/Microsoft/TypeScript/issues/1927), we use the tsconfig-glob npm package.

**Comment by bill-mybiz** [comment]
@egamma We still are going to look into changing to an exclude structure if possible and see if this has any effect. We have a client/server structure and it's possible that it was the only way to get what we needed back when that design decision was made.

**Comment by dbaeumer** [comment]
@bill-mybiz I am still unsure whether you are seeing this in TS or JS. Would you be willing to share your workspace? That would ease tracking this down?

**Comment by bill-mybiz** [comment]
@dbaeumer From the [link above](https://github.com/Microsoft/vscode/issues/3791#issuecomment-194300157):

> This is a TypeScript project and there are no source js files in my ts files glob.

I'm not sure what you mean by share my workspace. I've also already linked to [my project](https://github.com/ibgib/stackGib) above. 

My workspace settings file was empty. I have just yesterday added a workspace settings file that points to my local version of typescript to use string literal types. My user settings is as follows:

```
// Place your settings in this file to overwrite the default settings
{
    //------- Editor configuration

    "editor.fontSize": 9,
    "editor.quickSuggestions": false,
    "editor.tabSize": 4,
    "editor.insertSpaces": true,
    "editor.wrappingColumn": 90,
    "editor.wrappingIndent": "indent",
    "editor.suggestOnTriggerCharacters": false,
    "editor.renderWhitespace": true,

    //-------- Window configuration --------
    "window.reopenFolders": "all",

    //------- Diff Editor
    "diffEditor.renderSideBySide": false,

    //------- Files configuration

    "files.trimTrailingWhitespace": true,

    //------- Search configuration

    "search.exclude": {
        "**/bin": true,
        "**/bin-webpack": true,
        "**/node_modules": true,
        "**/platforms": true,
        "**/plugins": true,
        "**/typings": true
    },

    //------- Telemetry configuration

    "telemetry.enableCrashReporter": false,


    //-------- JavaScript configuration --------

    // Missing semicolon.
	"javascript.validate.lint.missingSemicolon": "warning"


    //-------- TypeScript configuration --------

	// Complete functions with their parameter signature.
	// "typescript.useCodeSnippetsOnMethodSuggest": true
}
```

A quick rundown of the structure of the [project](https://github.com/ibgib/stackGib): It uses a node server, an ionic/cordova client, and common shared code. It's a relatively small project. It uses webpack on both the client and the server. It uses gulp to compile what webpack doesn't, which is very little now. We still use DefinitelyTyped for type definitions. The code targets es5 and we are working on changing it to es6. 

Environment-wise, I am running on Windows 10 Pro in a Hyper-V guest VM with lots of ram and processing power with 6 dedicated cores.

**Comment by bill-mybiz** [comment]
I just got some similar lagged behavior after pointing to upgraded TS (1.8.7), so the issue may be related to TS and not VS code. Let me work through this this weekend and I'll post some more info.

**Comment by adamweeks** [comment]
I seem to be experiencing the same issue. Although, my application is pure JS, no TS. Hovering over anything in JS I get a tooltip of "Loading..." and nothing happens. 

I have two main projects that I use VS Code for and it only happens on one of them. The other loads fine. The main difference in the project with the issue is that it uses `nvm` for node versioning compatibility. I've made sure `tsc` can be run in both environments and that seems to be fine. 

Please let me know if I can provide any other info that might be useful.

**Comment by bill-mybiz** [comment]
@adamweeks If it's JS then it would be a different issue. Have you read through this [similar JavaScript issue](https://github.com/Microsoft/vscode/issues/3791)?

**Comment by adamweeks** [comment]
@bill-mybiz yep, ignore me :)

**Comment by xuyungit** [comment]
Any followup for this issue?
I found 0.10.10 works find when I switch the tsdk to the version in 0.10.9 (typescript 1.7.5),

**Comment by dbaeumer** [comment]
@bill-mybiz I cloned your repository and played with it but I didn't see any intellisense slowness. Can you provide some steps I can play with?

**Comment by bill-mybiz** [comment]
Well this weekend I was feeling under the weather and we didn't get as far as we wanted to. We did get to the point of compiling using excludes and not the ts-glob and that didn't seem to make any difference.

@dbaeumer Thank you for cloning and checking! [Here are some steps](https://github.com/Microsoft/vscode/issues/3958#issue-139891900) that you can try. If you get the trycatch snippet to pop up immediately, then you are not reproducing the lag.

I am thinking this morning that it might have something to do with our current version of node. We had to keep it downgraded because of previous conflicts with Ionic/Cordova as well as others, but it's possible that that those conflicts are now resolved. I'm going to try upgrading to some 4.+ version to see if that makes a difference (fingers crossed). If it doesn't, I'm going to double-check like @xuyungit and see if it's strictly the TS version, as I get a similar build lag (not intellisense) when I pointed my 0.10.9 version to TS 1.8+.

**Comment by bill-mybiz** [comment]
Well, I've upgraded to Node 4.2.4 and that has taken care of my build time problem, reducing the time from consistently ~110 seconds to ~20 seconds. So there must have been an issue with Typescript 1.8+ and Node 0.12.7 (previously used version) in my environment. There continues, however, to be lag in VS code. I'm still not 100% but I wanted to give this status update and respond. Thanks for the continued help!

**Comment by bill-mybiz** [comment]
I confirm what @xuyungit has said: 0.10.10 (and 0.10.11) works with TS 1.7.5 but not later. This includes both 1.8+ and `typescript@1.9.0-dev.20160128` build listed at [January's Release Notes](https://code.visualstudio.com/updates/vJanuary) under Enabling Salsa. So these do not work. Each time I try to use Intellisense, there is significant lag. The  easiest reproduction steps as I mentioned earlier involve the `trycatch` snippet. If you type out `trycatch` and hit ctrl+spacebar, you should get the "Try-Catch Statement" Intellisense under a second. Instead, it takes usually about ~10 seconds.

After further testing this morning, I have found that if I add my typings folder to the exclude statement in tsconfig that it works as expected. This might explain why @dbaeumer couldn't reproduce it if the typings weren't rebuilt.

**Comment by bill-mybiz** [comment]
Yes, so it definitely is related to the "Typings" folder. It appears that there isn't any caching being done on this folder, as when it is included the delay changes from <1s to ~10s. Apparently, whatever the process is that allows for intellisense, symbol renaming/parsing, etc. is not caching properly. I'm going to try to uninstall completely VS code and reinstall to see if that has any effect.

**Comment by bill-mybiz** [comment]
I've now uninstalled, deleted the Code folder in %appdata%, and reinstalled VS Code with still no luck.

**Comment by dbaeumer** [comment]
@bill-mybiz I was busy today doing a lot of other fixes but I will spent some time tomorrow trying to reproduce and understand this. A couple of comments:

- the node version installed on your system doesn't have an impact on the performance of the tsserver. The reason being that we use the 'electron' version we ship with (which is node as well) to run the tsserver 
- the tsserver very aggressive caches ts files and if resolved once keeps most of the info in memory.

**Comment by dbaeumer** [comment]
@bill-mybiz since the steps are easy I did some testing with the try catch snippet / code complete but I am not able to reproduce. Here is how my sytem behaves.

![bug](https://cloud.githubusercontent.com/assets/1931590/13859192/91709210-ec82-11e5-992a-96118d826498.gif)

Would you be able to install a process monitor tool on your system (like Process Hacker) and see which of the vscode processes have unusual CPU behavior in this scenario. If we know the process it will be easier to track this down However over the process which consumes the CPU and send me the command line of the process.

![capture](https://cloud.githubusercontent.com/assets/1931590/13859302/22950820-ec83-11e5-86fb-407b7725e910.png)

**Comment by bill-mybiz** [comment]
@dbaeumer As luck would have it, I already have Process Hacker installed. I have a picture, but GitHub doesn't allow me to upload them. Do you have another way for me to send you the pic?

**Comment by bill-mybiz** [comment]
![trycatchlag_processhacker](https://cloud.githubusercontent.com/assets/1253349/13863896/5daf9a64-ec6b-11e5-9ccb-0453c6fddb92.PNG)

**Comment by bill-mybiz** [comment]
That is with me specifying the typescript version in my local node_modules folder. The lag occurs also if I turn off that flag, but I'll go ahead and get screenshots of that too.

**Comment by bill-mybiz** [comment]
Ok, I was able to reproduce your behavior in that particular file. If I close VS code, open and do the particular changes you are making without touching the rest of the code, then it reproduces your behavior.

However, if I go to code that I'm working on right now in test/common/thing/thingexpression/thingexpress-tests.ts and add a trycatch inside of some non-trivial code (e.g. line 256), then it reproduces the lag. If I then go back to routes.ts that you chose (chats) and do the same edits, it will produce the lag.

**Comment by bill-mybiz** [comment]
I have just pushed the changes that get you at the exact same line that I'm working on, so you will need to pull again if you aren't able to get the behavior. You can also try it pretty much anywhere in one of those tests. For example, I get it when I try at line 149 just after `let logContext...`.

**Comment by bill-mybiz** [comment]
![trycatchlag_processhacker_2](https://cloud.githubusercontent.com/assets/1253349/13864219/ba41bf80-ec6d-11e5-8c97-9889a5fa9bd2.PNG)

**Comment by bill-mybiz** [comment]
And to show that it happens elsewhere, I decided to test elsewhere. I also got the behavior at common/thing/thing.expression.ts in the constructor. However, when I closed and restarted again and tried in the constructor again, it did not produce the lag. 

If I then go back into the thingexpress-tests.ts file and make a change there and return to the constructor, the lag is still there (and continues to be there until I restart VS code it appears after a little testing). So there is something about the thingexpress-tests.ts file (and possibly others) that is initiating the lag. 

Both of these files (thingexpress-tests.ts and thing.expression.ts) have import statements from rxjs and lodash, as well as multiple local file imports. The only substantive difference (insofar as the imports go) is that I have the chai imports at the top:
```
import chai = require('chai');
let assert = chai.assert;
let expect = chai.expect;
let should = chai.should();
```

Also, that tests file does have a lot of nested callbacks, as it is a mocha unit testing file and I use asynchronous rxjs (v5) quite a bit. So I'm thinking either the complexity of the structure is frying the TS brains or the chai import is mucking things up somehow. I would lean toward the complexity part but I think you will be able to reproduce at this point and have a better idea than me.

**Comment by bill-mybiz** [comment]
Ok, last comment for now: I tried just making changes to the thing.expression.ts constructor (line 54) and I am able to reproduce the lag if I just have the thingexpress-tests.ts open in the second window column (ctrl+2) *and don't even edit it*. If I just have it open, then the lag is produced while only editing in the thing.expression.ts constructor (line 54). The lag is also reproduced while making the edits you were making in chat routes.ts, as long as the thingexpress-tests.ts file is open in the second window.

Ok, that's it for me for now. Thanks for your help @dbaeumer.

**Comment by dbaeumer** [comment]
@bill-mybiz thanks for all your additional information. I tried to reproduce following your steps but wasn't able to do so. I will continue trying. The process explorer shows that the tsserver consumes quite some CPU so we know at least that there is something wrong with the server in our setup.

I am already on 1.8.9 which got released yesterday. If you have some time, could you give it a try on that version as well.

**Comment by bill-mybiz** [comment]
@dbaeumer 1.8.9 seems to have fixed the issue. I say this tentatively, as I haven't really tested it extensively. But if I use 1.8.2, the lag is immediate on restart of VS code in that thingexpress-tests.ts file. When I switch it to use 1.8.9, the lag doesn't seem to occur in that file. When I switch back, the lag returns. And for complete-ishness, I've now switched back to 1.8.9 and the lag is gone.

I did find that if I commented out most of those other unit/integration tests in that file, thus simplifying the code in that file, that the lag did not exist. If you did a pull on my code and you were on 1.8.2, that may be why you didn't reproduce the lag (I commented them out because I was just focusing on that last test). Since the complexity of the code seems to be causing the issue, it looks like perhaps 1.8.9 has some more performance enhancements that reduce the latency?

Either way, for now I would tentatively close this and if I have any problems I would reopen it. Does that sound good to you?

**Comment by xuyungit** [comment]
@bill-mybiz @dbaeumer I can also confirm that TS 1.8.9 also fixed the issue in my machine.

**Comment by bill-mybiz** [comment]
@xuyungit Thanks for the confirmation and your input. Much appreciated.
@dbaeumer Thanks for the help and work. Also much appreciated.

I also wanted to mention that my build time overall, outside of vs code is also much faster now using TS 1.8.9. 

I really appreciate the hard work put in by the TypeScript team and the VS Code team. Awesome stuff, both the product and the community.

**Comment by dbaeumer** [comment]
@bill-mybiz good to hear that the issue is fixed in 1.8.9. And thanks a lot for your detailed reports and your patience with this.
